### DISCUSSION TITLE
KeyEvent: when is text to be input?

### DISCUSSION THREAD

**Discussion by dhardy** [rust-windowing/winit#2997] [Q&A]
Context: I have a text editing widget which can both be typed into and can match (or rather a higher layer can match) sequences like `Ctrl+C`.

When my app receives a `KeyboardInput { event, .. }`, as I understand it `event.text` is the replacement for the old `ReceivedCharacter(c)`. *But*, when should text be input in the text editing widget?

- Only when `event.state == Pressed`, otherwise you get duplicate keys
- Not when `Control` is pressed ...

So, should the widget accept input when `event.state == ElementState::Pressed && self.modifiers.difference(ModifiersState::SHIFT).is_empty()` ? This appears to work, but I might be missing corner cases.

**Comment by kchibisov** [comment]
Pretty much yes? You can use though the `text_with_all_modifiers` and accept only printable chars, that will filter control keys automatically.

Yes, the input is usually from the press. Probably should also ignore `Dead`, because it's similar to what preedit is...

**Comment by dhardy** [comment]
Thanks. Didn't think about `Dead`.

Honestly, I'm not keen on the way `winit::platform::modifier_supplement::KeyEventExtModifierSupplement` is feature-gated; it just makes writing platform-portable code a chore. Anyway, sounds like it's just as easy not to use `text_with_all_modifiers`.

**Comment by kchibisov** [comment]
It's gated because it's different on mobile and you just get what you get sometimes and I think providing `None` is not a really good option, because it could be confusing?
